lib/recursive-methods.rb

```diff
@@ -1,15 +1,25 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  elsif n == 0 || n == 1
+    return 1
+  else
+    return n * (factorial(n - 1))
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s
+  else
+    return reverse(s[1..-1]) + s[0]
```

Comment:
`s[1..-1]` creates a new array and  copies all the individual elements over and so is O(n) by itself.